src/dj.erl

```diff
@@ -1,222 +1,810 @@
+%%
+%% ```
+%%    Dec = dj:one_of([ dj:field(online, dj:boolean())
+%%                    , dj:succeed(false)
+%%                    ]),
+%%    Json = << "[ {\"online\": true}"
+%%            , ", {\"online\": false}"
+%%            , ", {} ]"
+%%           >>,
+%%    {ok, [true, false, false]} = dj:decode(Json, dj:list(Dec)).
+%% '''
+%%
+%% This function is also useful for hardcoding values in {@link to_map/1},
+%% handling failure and success in {@link chain/2} and - as demonstrated -
+%% defaulting values using {@link one_of/1}.
+%%
+%% @see chain/2
+%% @see to_map/1
+%% @see fail/1
+%% @see one_of/1
+-spec succeed(T) -> decoder(T).
+succeed(V) ->
+  fun (_) ->
+      {ok, V}
+  end.
 
-compose(Fs) when is_list(Fs) ->
-  lists:foldl(fun compose/2, fun (X) -> X end, Fs).
+%% @doc Apply an n-ary function against a list of n decoders
+%%
+%% ```
+%%    Dec = dj:mapn( fun (X, Y, Z) -> {X, Y, Z} end
+%%                 , [ dj:field(major, dj:pos_integer())
+%%                   , dj:field(minor, dj:non_neg_integer())
+%%                   , dj:field(patch, dj:non_neg_integer())
+%%                   ]
+%%                 ),
+%%    Json = << "{ \"major\": 123"
+%%            , ", \"minor\": 66"
+%%            , ", \"patch\": 0"
+%%            , "}">>,
+%%    {ok, {123, 66, 0}} = dj:decode(Json, Dec).
+%% '''
+%%
+%% @see map/2
+%% @see chain/2
+-spec mapn(Fun, [decoder(T)]) -> decoder(V) when
+    Fun :: function(),
+    T   :: term(),
+    V   :: term().
+mapn(Fun, Decoders) when is_function(Fun, length(Decoders))->
+  map(fun (Vs) -> erlang:apply(Fun, Vs) end, sequence(Decoders)).
 
-compose(F, G) ->
-  fun (X) -> F(G(X)) end.
+%% @doc Decoding succeeds if the decoder produces exactly the supplied value.
+%%
+%% This can, for example, be used when a certain field is used to switch between
+%% different decoders.
+-spec exactly(V, decoder(V)) -> decoder(V).
+exactly(V, Decoder) ->
+  fun (Json) ->
+      case Decoder(Json) of
+        {ok, V} -> {ok, V};
+        _       -> {error, [{not_exactly, V, Json}]}
+      end
+  end.
 
-sequence(Fs, X) ->
-  sequence(Fs, X, []).
+%% @doc Try a bunch of decoders. The first one to succeed will be used.
+%%
+%% If all decoders fails, the errors are accumulated.
+%%
+%% ```
+%%    Dec = dj:one_of([dj:binary(), dj:integer()]),
+%%    {ok, <<"foo">>} = dj:decode(<<"\"foo\"">>, Dec),
+%%    {ok, 123} = dj:decode(<<"123">>, Dec),
+%%    {error, _} = dj:decode(<<"null">>, Dec).
+%% '''
+-spec one_of([decoder(V), ...]) -> decoder(V).
+one_of(Decoders) ->
+  fun (Json) ->
+      try_decoders(Decoders, Json, [])
+  end.
 
-sequence([], _, Ys) ->
-  lists:reverse(Ys);
-sequence([F | Fs], X, Ys) ->
-  sequence(Fs, X, [F(X) | Ys]).
+%% @equiv dj:decode(Json, Decoder, [{labels, existing_atom}])
+-spec decode(Json, decoder(T)) -> result(T, errors()) when
+    Json :: jsx:json_text().
+decode(Json, Decoder) ->
+  decode(Json, Decoder, [{labels, existing_atom}]).
```

Comment:
@sgillis @kwrooijen I've been considering changing this to `attempt_atom`.

So the main reason for using `existing_atom` here is to minimize resource exhaustion attack surface, since atoms are never garbage collected.

However, this also means that `dj:decode/2` is incapable of decoding json that contains a key which is not a known atom.

By using `attempt_atom` instead, you could still _use_ `dj` as if it were using `existing_atom`: by only passing atoms to `dj:field/2` and `dj:at/2`, you create the atoms, hence ensuring they exist!

_If_ any unknown keys do exist, they'd simply stay binaries.

Wdyt? Are there downsides to such an approach?

Reply:
IMHO, I think we should filter out fields that have not been specified in the DJ decoder. Why should we receive extra data we were not expecting?

Reply:
To be fair, the only way that I can think of to actually access those fields is using `dj:value/0` which ostensibly returns the `jsx:json_term()` from the JSON.

Note that "data I don't need" doesn't necessarily mean "data I don't expect". When you have full control over the source of the JSON, you can indeed make sure it doesn't send any fields you're not interested in. When dealing with external API's however, you don't always have that control.

It's important here to remark that using decoders other than `dj:value/0` does filter out any data you don't need.

Are you suggesting to make `dj:decode/2` and `dj:decode/3` fail (or warn) when not all data in the json is actually used? ([prior art, in Elm](http://package.elm-lang.org/packages/zwilias/json-decode-exploration/5.0.0/Json-Decode-Exploration))

Reply:
As discussed in private: I agree with `attempt_atom`